### PR TITLE
8213714: AttachingConnector/attach/attach001 failed due to "bind failed: Address already in use"

### PR DESCRIPTION
**Pull request by alexmenkov**
The fix updates the tests to use dynamic port launching debuggee and get the listening port from the debugee output

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8213714](https://bugs.openjdk.java.net/browse/JDK-8213714): AttachingConnector/attach/attach001 failed due to "bind failed: Address already in use"

### Reviewers
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4817/head:pull/4817` \
`$ git checkout pull/4817`

Update a local copy of the PR: \
`$ git checkout pull/4817` \
`$ git pull https://git.openjdk.java.net/jdk pull/4817/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4817`

View PR using the GUI difftool: \
`$ git pr show -t 4817`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4817.diff">https://git.openjdk.java.net/jdk/pull/4817.diff</a>

</details>

### CODE REVIEW

**Comment by alexmenkov** [comment]
Ping

File: test/hotspot/jtreg/vmTestbase/nsk/share/IORedirector.java [hunk 1]
```diff
@@ -67,16 +70,16 @@ public IORedirector(InputStream in, OutputStream out) {
      */
     public IORedirector(BufferedReader in, Log log, String prefix) {
         this();
-        this.prefix = prefix;
         this.bin  = in;
-        this.log = log;
+        addProcessor( s -> log.println(prefix + s));
```
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/share/IORedirector.java:74]: Nit: unneeded space before "s ->".

**Review by sspitsyn — CHANGES_REQUESTED** [review]
Hi Alex,

The fix looks good to me. While you are a this code, may I ask you to do a minor cleanup and get rid of several commented out lines that were there before your fix?

Thanks,
Serguei

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java [hunk 2]
```diff
@@ -326,12 +327,19 @@ public void redirectStdout(OutputStream out) {
      * was already redirected, the TestBug exception is thrown.
      */
     public void redirectStdout(Log log, String prefix) {
+        redirectStdout(log, prefix, null);
```
Comment (plummercj) [inline test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java:330]: Is there a reason to still support passing in the prefix even though it is ignored. Seems that could be confusing to someone looking at the caller code.

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/AttachingConnector/attachnosuspend/attachnosuspend001.java [hunk 3]
```diff
@@ -96,19 +94,17 @@ private int runIt(String argv[], PrintStream out) {
...
-        listenPort = argHandler.getTransportPort();
-
         String java = argHandler.getLaunchExecPath()
                         + " " + argHandler.getLaunchOptions();
-        String cmd = java +
-                " -Xdebug -Xnoagent -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=" +
-                listenPort + " " + DEBUGEE_CLASS;
+        String cmd = java
+                + " -Xdebug -Xnoagent -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=0"
+                + " " + DEBUGEE_CLASS;
```
Comment (plummercj) [inline test/hotspot/jtreg/vmTestbase/nsk/jdi/AttachingConnector/attachnosuspend/attachnosuspend001.java:101]: You've lost logging of the port you end up listening on. Is the dynamically discovered listening port logged somewhere else?

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java (hunk 2 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java:327]: I dont' quite follow.
The prefix is not ignored, it's passed to IORedirector and is used for logging

File: test/hotspot/jtreg/vmTestbase/nsk/jdi/AttachingConnector/attachnosuspend/attachnosuspend001.java (hunk 3 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/vmTestbase/nsk/jdi/AttachingConnector/attachnosuspend/attachnosuspend001.java:101]: DebugeeProcess.redirectOutputAndDetectListeningAddress logs it
(see DebugeeProcess.java, line 428)
Also IORedirector logs debugee output (including "Listening for transport ... at address ..." line)

File: test/hotspot/jtreg/vmTestbase/nsk/share/IORedirector.java (hunk 1 above)
Comment (alexmenkov) [reply test/hotspot/jtreg/vmTestbase/nsk/share/IORedirector.java]: Fixed

**Review by alexmenkov — COMMENTED** [review]
> The fix looks good to me. While you are a this code, may I ask you to do a minor cleanup and get rid of several commented out lines that were there before your fix?

Dropped commented lines in DebugeeProcess.java.
Also updated methods description as they were incorrect.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java (hunk 2 above)
Comment (plummercj) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java:327]: Sorry, I misread it. I thought the 3rd argument passed was the prefix.

**Review by plummercj — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]

**Pull request closed without merging**